### ISSUE TITLE
vcs: gix-blame intermittent object-lookup failure on pathologically repetitive files

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#579]
Discovered while implementing #329 (per-function VCS via `git blame`).

## Symptom

`gix::Repository::blame_file` (gix 0.83 → gix-blame 0.13) intermittently
returns a spurious error on files with a deep history (~130+ commits)
whose content is **pathologically repetitive** (e.g. thousands of
near-identical lines):

```
Could not find existing iterator over a tree
An object with id <sha> could not be found
```

The failure is **non-deterministic for an identical repository** — the
same fixture (same content, same fixed commit timestamps → identical
object SHAs) returns `Ok` on one run and `Err` on the next.

## Root cause (investigated 2026-06-08)

The defect is **not in gix-blame** — it is one layer below, in the
lock-free `gix-odb` pack-refresh path. Both error strings are plain
"expected object was reported missing" errors bubbling up from the object
store through `gix_blame::Error::{FindExistingObject, FindExistingIter}`.
gix-blame holds no object cache of its own and does no retry. The
non-deterministic-on-identical-input symptom matches the known ODB race in
[gitoxide discussion #1412](https://github.com/GitoxideLabs/gitoxide/discussions/1412):
under concurrent ODB access a thread can momentarily fail to observe a
freshly-loaded pack index and conclude an object is missing. Whole-repo
analysis blames from a worker pool sharing one `ThreadSafeRepository` ODB
— exactly the contended case. Deep history + repetitive content drives
many concurrent lookups, which is why only the degenerate stress fixture
trips it.

**Bumping does not fix it:** still present in the latest upstream release
(gix 0.84 / gix-blame 0.14); the 0.13→0.14 changelog has no object-lookup
/ caching / correctness fix. (Locally, the failure could not be reproduced
even at 10k identical lines / 200 commits across 320 concurrent trials —
it is load-/timing-/host-dependent.)

## Real-world impact: low

Blame is rock-solid on real source: this repository's own `CHANGELOG.md`
(217 commits) blames cleanly, and a synthetic 10k-line / 200-commit file
with *unique* per-line content blames in ~100ms with no errors. Only the
degenerate "identical lines" stress fixture trips it.

## Fix shipped on our side (mitigation)

Two defences now apply in `src/vcs/git/blame.rs`:

1. **Bounded retry** of every ODB object lookup that can hit the transient
   miss — the whole-file blame **and** the post-blame `find_commit`
   resolution — up to `MAX_BLAME_ATTEMPTS` (3) times. Each retry re-reads
   the index snapshot, turning a momentary miss into a hit. The retryable
   set is matched on named gix error variants (`FindExistingObject` /
   `FindExistingIter` for blame; nested `NotFound` for the commit lookup),
   so a deterministic error (file missing at the ref, object genuinely
   absent, decode/conversion failure) is never retried, and an upstream
   variant rename breaks our build rather than silently degrading.
2. **Graceful degradation** (pre-existing, #329): if every retry still
   fails, the front end skips that one file's per-function blocks and keeps
   the file-level `vcs` block, so an unblameable file never aborts the walk.

A generic, unit-tested `retry_transient` helper carries the retry logic
(mutation-verified: disabling retry, ignoring the predicate, and an
off-by-one budget each break a test).

## Action items

- [ ] **Report upstream** against `GitoxideLabs/gitoxide` (gix-blame),
      framed as an ODB-level non-deterministic lookup race surfaced via
      `blame_file`, cross-referencing discussion #1412. A ready-to-paste
      draft (summary + analysis + reproducer skeleton) is in the
      collapsible block below.
- [x] **Re-test on a gix bump** — done: gix 0.84 / gix-blame 0.14 does not
      fix it.

<details>
<summary>Ready-to-paste upstream issue draft</summary>

# `blame_file` intermittently reports a spurious "object could not be found" under concurrent ODB access

## Summary

`gix::Repository::blame_file` (gix 0.83 / gix-blame 0.13) **intermittently**
returns one of:

```
Could not find existing iterator over a tree
An object with id <sha> could not be found
```

on a file with deep history (~130+ commits) whose content is pathologically
repetitive (thousands of near-identical lines). The failure is
**non-deterministic for a byte-identical repository** (fixed commit
timestamps → identical object SHAs): the same fixture returns `Ok` on one run
and `Err` on the next.

## Why this looks like an ODB-level race, not a blame bug

Both error strings are plain "expected object was reported missing" errors
bubbling up from `gix-odb` through `gix_blame::Error::{FindExistingObject,
FindExistingIter}`. gix-blame holds no object cache of its own and does no
retry — it relies entirely on the shared object store. The symptom (sporadic,
non-deterministic "object could not be found" on objects that provably exist,
surfacing only under many concurrent lookups) matches the lock-free
ODB/pack-refresh behaviour discussed in
https://github.com/GitoxideLabs/gitoxide/discussions/1412 : a thread can
momentarily fail to observe a freshly-loaded pack index and conclude an object
is missing.

It surfaces only when the same `ThreadSafeRepository` ODB is hit from many
threads at once (e.g. a worker pool blaming many files), and only on
deep-history + repetitive content (many object lookups + heavy diff churn →
lots of concurrent ODB hits). It does **not** reproduce on real source.

## Status / versions

Still present in the latest release as of 2026-06-08 (gix 0.84 / gix-blame
0.14). gix-blame's 0.13→0.14 changelog has no object-lookup / caching /
correctness fix.

## Reproducer (load-dependent — may need many iterations / a contended host)

Build a repo of N identical lines wrapped in one function, make ~150-200
commits each touching one rotating line, then blame the whole file repeatedly
from a pool of threads sharing one `ThreadSafeRepository`. On an uncontended
machine it may take many iterations to trip; raising the thread count, line
count (10k+), and history depth increases the hit rate. Skeleton:

```rust
// pseudocode — share one ThreadSafeRepository across threads
let repo = gix::open(path)?.into_sync();
let handles: Vec<_> = (0..N_THREADS).map(|_| {
    let repo = repo.clone();
    std::thread::spawn(move || {
        let repo = repo.to_thread_local();
        for _ in 0..ITERS {
            let _ = repo.blame_file(file.as_ref(), head, opts.clone());
            // intermittently Err(FindExistingObject|FindExistingIter)
        }
    })
}).collect();
```

## Downstream mitigation (for reference)

In big-code-analysis we now retry the transient lookup (blame + the post-blame
`find_commit`) a bounded number of times, which turns the momentary miss into
a hit; if it still fails we degrade gracefully (skip that file's per-function
blocks). Tracking issue: GitoxideLabs downstream consumer dekobon/big-code-analysis#579.

</details>

---

Not closing: the upstream report (action item 1) is left for a maintainer
to file. The downstream mitigation is complete (branch
`fix/579-blame-odb-lookup-retry`).
